## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#43990c24-5f30-4fb6-bf96-6769c2e24cf8) |
| mint-fs.sh | :x: [more...](#a8bf12ef-fa15-406b-a10b-7973f86ff112) |
| mint-gateway-s3.sh | :x: [more...](#e8d57906-e7bb-4733-b2cd-9a17b66d3a12) |
| mint-gateway-azure.sh | :x: [more...](#1f4d0078-0ef1-43ca-8d75-60245f048835) |
<a name="43990c24-5f30-4fb6-bf96-6769c2e24cf8"></a>
### 9616-ad57acd/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31861
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e7b55a6175c1:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... FAILED in 11 minutes and 16 seconds
{"name": "awscli", "duration": 629824, "function": "aws --endpoint-url http://minio-dev4.minio.io:31861 s3 cp /mint/data/datafile-65-MB s3://awscli-mint-test-bucket-18680/datafile-65-MB\n", "status": "FAIL", "error": "Completed 256.0 KiB/65.0 MiB (3.6 MiB/s) with 1 file(s) remaining
Completed 512.0 KiB/65.0 MiB (6.9 MiB/s) with 1 file(s) remaining
Completed 768.0 KiB/65.0 MiB (9.8 MiB/s) with 1 file(s) remaining
Completed 1.0 MiB/65.0 MiB (12.7 MiB/s) with 1 file(s) remaining 
Completed 1.2 MiB/65.0 MiB (10.0 MiB/s) with 1 file(s) remaining 
Completed 1.5 MiB/65.0 MiB (11.9 MiB/s) with 1 file(s) remaining 
Completed 1.8 MiB/65.0 MiB (13.2 MiB/s) with 1 file(s) remaining 
Completed 2.0 MiB/65.0 MiB (14.9 MiB/s) with 1 file(s) remaining 
Completed 2.2 MiB/65.0 MiB (16.6 MiB/s) with 1 file(s) remaining 
Completed 2.5 MiB/65.0 MiB (17.5 MiB/s) with 1 file(s) remaining 
Completed 2.8 MiB/65.0 MiB (19.1 MiB/s) with 1 file(s) remaining 
Completed 3.0 MiB/65.0 MiB (20.7 MiB/s) with 1 file(s) remaining 
Completed 3.2 MiB/65.0 MiB (22.0 MiB/s) with 1 file(s) remaining 
Completed 3.5 MiB/65.0 MiB (22.7 MiB/s) with 1 file(s) remaining 
Completed 3.8 MiB/65.0 MiB (24.0 MiB/s) with 1 file(s) remaining 
Completed 4.0 MiB/65.0 MiB (25.1 MiB/s) with 1 file(s) remaining 
Completed 4.2 MiB/65.0 MiB (26.4 MiB/s) with 1 file(s) remaining 
Completed 4.5 MiB/65.0 MiB (27.6 MiB/s) with 1 file(s) remaining 
Completed 4.8 MiB/65.0 MiB (28.7 MiB/s) with 1 file(s) remaining 
Completed 5.0 MiB/65.0 MiB (29.9 MiB/s) with 1 file(s) remaining 
Completed 5.2 MiB/65.0 MiB (31.0 MiB/s) with 1 file(s) remaining 
Completed 5.5 MiB/65.0 MiB (32.0 MiB/s) with 1 file(s) remaining 
Completed 5.8 MiB/65.0 MiB (33.0 MiB/s) with 1 file(s) remaining 
Completed 6.0 MiB/65.0 MiB (34.1 MiB/s) with 1 file(s) remaining 
Completed 6.2 MiB/65.0 MiB (34.9 MiB/s) with 1 file(s) remaining 
Completed 6.5 MiB/65.0 MiB (36.0 MiB/s) with 1 file(s) remaining 
Completed 6.8 MiB/65.0 MiB (36.8 MiB/s) with 1 file(s) remaining 
Completed 7.0 MiB/65.0 MiB (37.7 MiB/s) with 1 file(s) remaining 
Completed 7.2 MiB/65.0 MiB (38.4 MiB/s) with 1 file(s) remaining 
Completed 7.5 MiB/65.0 MiB (39.2 MiB/s) with 1 file(s) remaining 
Completed 7.8 MiB/65.0 MiB (40.3 MiB/s) with 1 file(s) remaining 
Completed 8.0 MiB/65.0 MiB (41.0 MiB/s) with 1 file(s) remaining 
Completed 8.2 MiB/65.0 MiB (41.6 MiB/s) with 1 file(s) remaining 
Completed 8.5 MiB/65.0 MiB (42.2 MiB/s) with 1 file(s) remaining 
Completed 8.8 MiB/65.0 MiB (43.2 MiB/s) with 1 file(s) remaining 
Completed 9.0 MiB/65.0 MiB (43.9 MiB/s) with 1 file(s) remaining 
Completed 9.2 MiB/65.0 MiB (8.2 MiB/s) with 1 file(s) remaining  
Completed 9.5 MiB/65.0 MiB (8.4 MiB/s) with 1 file(s) remaining  
Completed 9.8 MiB/65.0 MiB (8.6 MiB/s) with 1 file(s) remaining  
Completed 10.0 MiB/65.0 MiB (8.7 MiB/s) with 1 file(s) remaining 
Completed 10.2 MiB/65.0 MiB (8.9 MiB/s) with 1 file(s) remaining 
Completed 10.5 MiB/65.0 MiB (9.1 MiB/s) with 1 file(s) remaining 
Completed 10.8 MiB/65.0 MiB (9.3 MiB/s) with 1 file(s) remaining 
Completed 11.0 MiB/65.0 MiB (9.5 MiB/s) with 1 file(s) remaining 
Completed 11.2 MiB/65.0 MiB (9.6 MiB/s) with 1 file(s) remaining 
Completed 11.5 MiB/65.0 MiB (9.8 MiB/s) with 1 file(s) remaining 
Completed 11.8 MiB/65.0 MiB (9.9 MiB/s) with 1 file(s) remaining 
Completed 12.0 MiB/65.0 MiB (10.1 MiB/s) with 1 file(s) remaining
Completed 12.2 MiB/65.0 MiB (10.3 MiB/s) with 1 file(s) remaining
Completed 12.5 MiB/65.0 MiB (10.4 MiB/s) with 1 file(s) remaining
Completed 12.5 MiB/65.0 MiB (41.1 KiB/s) with 1 file(s) remaining
Completed 12.6 MiB/65.0 MiB (41.2 KiB/s) with 1 file(s) remaining
Completed 12.6 MiB/65.0 MiB (41.3 KiB/s) with 1 file(s) remaining
Completed 12.7 MiB/65.0 MiB (41.3 KiB/s) with 1 file(s) remaining
Completed 12.7 MiB/65.0 MiB (41.4 KiB/s) with 1 file(s) remaining
Completed 12.7 MiB/65.0 MiB (41.3 KiB/s) with 1 file(s) remaining
Completed 12.8 MiB/65.0 MiB (41.2 KiB/s) with 1 file(s) remaining
upload failed: ../../../data/datafile-65-MB to s3://awscli-mint-test-bucket-18680/datafile-65-MB Connection was closed before we received a valid response from endpoint URL: "http://minio-dev4.minio.io:31861/awscli-mint-test-bucket-18680/datafile-65-MB?uploadId=ac5400fa-6d9e-45b4-97c6-cf7a3aa5d0c8&partNumber=1"."}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... 
Aborting Mint...
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31861
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e7b55a6175c1:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 minutes and 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... FAILED in 11 minutes and 16 seconds
{"name": "awscli", "duration": 625323, "function": "aws --endpoint-url http://minio-dev4.minio.io:31861 s3 cp /mint/data/datafile-65-MB s3://awscli-mint-test-bucket-8600/datafile-65-MB\n", "status": "FAIL", "error": "Completed 256.0 KiB/65.0 MiB (3.5 MiB/s) with 1 file(s) remaining
Completed 512.0 KiB/65.0 MiB (6.5 MiB/s) with 1 file(s) remaining
Completed 768.0 KiB/65.0 MiB (8.8 MiB/s) with 1 file(s) remaining
Completed 1.0 MiB/65.0 MiB (11.5 MiB/s) with 1 file(s) remaining 
Completed 1.2 MiB/65.0 MiB (9.2 MiB/s) with 1 file(s) remaining  
Completed 1.5 MiB/65.0 MiB (10.9 MiB/s) with 1 file(s) remaining 
Completed 1.8 MiB/65.0 MiB (12.4 MiB/s) with 1 file(s) remaining 
Completed 2.0 MiB/65.0 MiB (13.8 MiB/s) with 1 file(s) remaining 
Completed 2.2 MiB/65.0 MiB (15.1 MiB/s) with 1 file(s) remaining 
Completed 2.5 MiB/65.0 MiB (16.5 MiB/s) with 1 file(s) remaining 
Completed 2.8 MiB/65.0 MiB (17.7 MiB/s) with 1 file(s) remaining 
Completed 3.0 MiB/65.0 MiB (18.9 MiB/s) with 1 file(s) remaining 
Completed 3.2 MiB/65.0 MiB (20.1 MiB/s) with 1 file(s) remaining 
Completed 3.5 MiB/65.0 MiB (21.2 MiB/s) with 1 file(s) remaining 
Completed 3.8 MiB/65.0 MiB (22.3 MiB/s) with 1 file(s) remaining 
Completed 4.0 MiB/65.0 MiB (23.5 MiB/s) with 1 file(s) remaining 
Completed 4.2 MiB/65.0 MiB (24.7 MiB/s) with 1 file(s) remaining 
Completed 4.5 MiB/65.0 MiB (25.7 MiB/s) with 1 file(s) remaining 
Completed 4.8 MiB/65.0 MiB (26.6 MiB/s) with 1 file(s) remaining 
Completed 5.0 MiB/65.0 MiB (27.6 MiB/s) with 1 file(s) remaining 
Completed 5.2 MiB/65.0 MiB (28.5 MiB/s) with 1 file(s) remaining 
Completed 5.5 MiB/65.0 MiB (29.4 MiB/s) with 1 file(s) remaining 
Completed 5.8 MiB/65.0 MiB (30.4 MiB/s) with 1 file(s) remaining 
Completed 6.0 MiB/65.0 MiB (31.3 MiB/s) with 1 file(s) remaining 
Completed 6.2 MiB/65.0 MiB (32.3 MiB/s) with 1 file(s) remaining 
Completed 6.5 MiB/65.0 MiB (33.3 MiB/s) with 1 file(s) remaining 
Completed 6.8 MiB/65.0 MiB (33.9 MiB/s) with 1 file(s) remaining 
Completed 7.0 MiB/65.0 MiB (34.7 MiB/s) with 1 file(s) remaining 
Completed 7.2 MiB/65.0 MiB (35.6 MiB/s) with 1 file(s) remaining 
Completed 7.5 MiB/65.0 MiB (36.5 MiB/s) with 1 file(s) remaining 
Completed 7.8 MiB/65.0 MiB (37.2 MiB/s) with 1 file(s) remaining 
Completed 8.0 MiB/65.0 MiB (38.1 MiB/s) with 1 file(s) remaining 
Completed 8.2 MiB/65.0 MiB (38.9 MiB/s) with 1 file(s) remaining 
Completed 8.5 MiB/65.0 MiB (39.6 MiB/s) with 1 file(s) remaining 
Completed 8.8 MiB/65.0 MiB (40.5 MiB/s) with 1 file(s) remaining 
Completed 9.0 MiB/65.0 MiB (41.2 MiB/s) with 1 file(s) remaining 
Completed 9.2 MiB/65.0 MiB (8.1 MiB/s) with 1 file(s) remaining  
Completed 9.5 MiB/65.0 MiB (8.3 MiB/s) with 1 file(s) remaining  
Completed 9.8 MiB/65.0 MiB (8.5 MiB/s) with 1 file(s) remaining  
Completed 10.0 MiB/65.0 MiB (8.7 MiB/s) with 1 file(s) remaining 
Completed 10.2 MiB/65.0 MiB (8.9 MiB/s) with 1 file(s) remaining 
Completed 10.5 MiB/65.0 MiB (9.1 MiB/s) with 1 file(s) remaining 
Completed 10.8 MiB/65.0 MiB (9.3 MiB/s) with 1 file(s) remaining 
Completed 10.9 MiB/65.0 MiB (35.9 KiB/s) with 1 file(s) remaining
Completed 11.1 MiB/65.0 MiB (36.4 KiB/s) with 1 file(s) remaining
Completed 11.3 MiB/65.0 MiB (36.9 KiB/s) with 1 file(s) remaining
Completed 11.4 MiB/65.0 MiB (37.3 KiB/s) with 1 file(s) remaining
Completed 11.6 MiB/65.0 MiB (37.8 KiB/s) with 1 file(s) remaining
Completed 11.8 MiB/65.0 MiB (38.3 KiB/s) with 1 file(s) remaining
Completed 12.0 MiB/65.0 MiB (38.8 KiB/s) with 1 file(s) remaining
upload failed: ../../../data/datafile-65-MB to s3://awscli-mint-test-bucket-8600/datafile-65-MB Connection was closed before we received a valid response from endpoint URL: "http://minio-dev4.minio.io:31861/awscli-mint-test-bucket-8600/datafile-65-MB?uploadId=d3d749d3-7fe1-4edb-ae11-3f3bb3323508&partNumber=3"."}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... 
Aborting Mint...
```
<a name="1f4d0078-0ef1-43ca-8d75-60245f048835"></a>
### 9616-ad57acd/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31773
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 5174fdc4dc1b:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 1 minutes and 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 14 seconds
(5/15) Running awscli tests ... done in 2 minutes and 47 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 2 seconds
(8/15) Running minio-dotnet tests ... done in 2 minutes and 26 seconds
(9/15) Running minio-go tests ... done in 3 minutes and 41 seconds
(10/15) Running minio-java tests ... done in 27 minutes and 46 seconds
(11/15) Running minio-js tests ... done in 2 minutes and 45 seconds
(12/15) Running minio-py tests ... FAILED in 49 minutes and 19 seconds
{
  "name": "minio-py:test_thread_safe",
  "function": "put_object(bucket_name, object_name, data, length, content_type, metadata, sse, progress, part_size)",
  "args": {
    "bucket_name": "minio-py-test-5e45a855-7500-4e93-8a35-8f81dfb26daa",
    "object_name": "0f0a8be3-d6c0-4d71-a946-dc673a595276"
  },
  "duration": 2403344,
  "message": "RequestTimeTooSkewed: message: The difference between the request time and the server's time is too large.",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1638, in test_thread_safe\n    raise exceptions[0]\nException: RequestTimeTooSkewed: message: The difference between the request time and the server's time is too large.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2128, in main\n    test_thread_safe(client, testfile, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1640, in test_thread_safe\n    raise Exception(err)\nException: RequestTimeTooSkewed: message: The difference between the request time and the server's time is too large.\n",
  "status": "FAIL"
}
(12/15) Running s3cmd tests ... done in 1 minutes and 25 seconds
(13/15) Running s3select tests ... done in 22 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="e8d57906-e7bb-4733-b2cd-9a17b66d3a12"></a>
### 9616-ad57acd/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32617
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 86ab382be64a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 5 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 26 seconds
(8/15) Running minio-dotnet tests ... done in 36 minutes and 7 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 1 seconds
{
  "args": {
    "bucketName": "minio-go-test-ui6gcva5092bcear",
    "region": "eu-west-1"
  },
  "duration": 60135,
  "error": "Your previous request to create the named bucket succeeded and you already own it.",
  "function": "MakeBucket(bucketName, region)",
  "message": "MakeBucket failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... FAILED in 1 minutes and 1 seconds
{
  "name": "minio-java",
  "function": "makeBucket(String bucketName)",
  "duration": 60310,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = BucketAlreadyOwnedByYou, message = Your previous request to create the named bucket succeeded and you already own it., bucketName = minio-java-test-erlhit, objectName = null, resource = /minio-java-test-erlhit, requestId = 160FB232587D8B3C, hostId = 8edbfde4-47e0-4f8d-8c32-39de8e1cef1d)\nrequest={method=PUT, url=http://minio-dev4.minio.io:32617/minio-java-test-erlhit, headers=Host: minio-dev4.minio.io:32617\nAccept-Encoding: identity\nUser-Agent: MinIO (amd64; amd64) minio-java/dev\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20200517T031202Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200517/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=409, headers=Accept-Ranges: bytes\nContent-Length: 374\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO/DEVELOPMENT.GOGET\nVary: Origin\nX-Amz-Request-Id: 160FB232587D8B3C\nX-Xss-Protection: 1; mode=block\nDate: Sun, 17 May 2020 03:13:02 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:1211), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.makeBucket(MinioClient.java:3286), io.minio.MinioClient.makeBucket(MinioClient.java:3199), FunctionalTest.makeBucket_test1(FunctionalTest.java:247), FunctionalTest.runTests(FunctionalTest.java:4005), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... FAILED in 11 minutes and 56 seconds
{
  "name": "minio-js",
  "function": "\"before all\" hook in \"functional tests\"",
  "duration": 60121,
  "status": "FAIL",
  "error": "S3Error: Your previous request to create the named bucket succeeded and you already own it. at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(9/15) Running minio-py tests ... done in 56 seconds
(10/15) Running s3cmd tests ... done in 1 minutes and 0 seconds
(11/15) Running s3select tests ... done in 1 seconds
(12/15) Running security tests ... done in 0 seconds

Executed 12 out of 15 tests successfully.
```
<a name="a8bf12ef-fa15-406b-a10b-7973f86ff112"></a>
### 9616-ad57acd/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32045
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp e97b5bc081bd:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 40 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 4 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 24 seconds
(8/15) Running minio-dotnet tests ... done in 1 hours and 10 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 0 seconds
{
  "args": {
    "bucketName": "minio-go-test-jn3dex0r0nex2wrn",
    "region": "eu-west-1"
  },
  "duration": 60110,
  "error": "Your previous request to create the named bucket succeeded and you already own it.",
  "function": "MakeBucket(bucketName, region)",
  "message": "MakeBucket failed",
  "name": "minio-go: testGetObjectClosedTwiceV2",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... FAILED in 15 minutes and 1 seconds
{
  "name": "minio-java",
  "function": "makeBucket(String bucketName)",
  "duration": 900106,
  "status": "FAIL",
  "error": "java.net.SocketTimeoutException: timeout >>> [okio.Okio$4.newTimeoutException(Okio.java:232), okio.AsyncTimeout.exit(AsyncTimeout.java:286), okio.AsyncTimeout$2.read(AsyncTimeout.java:241), okio.RealBufferedSource.indexOf(RealBufferedSource.java:358), okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:230), okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:226), okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:202), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:101), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:264), okhttp3.RealCall.execute(RealCall.java:93), io.minio.MinioClient.execute(MinioClient.java:1097), io.minio.MinioClient.execute(MinioClient.java:1236), io.minio.MinioClient.executePut(MinioClient.java:1410), io.minio.MinioClient.makeBucket(MinioClient.java:3286), io.minio.MinioClient.makeBucket(MinioClient.java:3199), FunctionalTest.makeBucket_test1(FunctionalTest.java:247), FunctionalTest.runTests(FunctionalTest.java:4005), FunctionalTest.main(FunctionalTest.java:4192)]"
}
(9/15) Running minio-js tests ... done in 1 minutes and 48 seconds
(10/15) Running minio-py tests ... done in 49 seconds
(11/15) Running s3cmd tests ... done in 14 seconds
(12/15) Running s3select tests ... done in 0 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/namespace-lock.go
+++ b/cmd/namespace-lock.go
@@ -18,17 +18,18 @@ package cmd
 
 import (
 	"context"
+	"errors"
 	pathutil "path"
 	"runtime"
 	"sort"
 	"strings"
 	"sync"
-	"sync/atomic"
 
 	"fmt"
 	"time"
 
 	"github.com/minio/lsync"
+	"github.com/minio/minio/cmd/logger"
 	"github.com/minio/minio/pkg/dsync"
 )
 
@@ -57,7 +58,7 @@ func newNSLock(isDistXL bool) *nsLockMap {
 
 // nsLock - provides primitives for locking critical namespace regions.
 type nsLock struct {
-	ref uint32
+	ref int32
 	*lsync.LRWMutex
 }
 
@@ -72,21 +73,18 @@ type nsLockMap struct {
 
 // Lock the namespace resource.
 func (n *nsLockMap) lock(ctx context.Context, volume string, path string, lockSource, opsID string, readLock bool, timeout time.Duration) (locked bool) {
-	var nsLk *nsLock
-
 	resource := pathJoin(volume, path)
 
 	n.lockMapMutex.Lock()
-	if _, found := n.lockMap[resource]; !found {
-		n.lockMap[resource] = &nsLock{
+	nsLk, found := n.lockMap[resource]
+	if !found {
+		nsLk = &nsLock{
 			LRWMutex: lsync.NewLRWMutex(ctx),
-			ref:      1,
 		}
-	} else {
-		// Update ref count here to avoid multiple races.
-		atomic.AddUint32(&n.lockMap[resource].ref, 1)
+		// Add a count to indicate that a parallel unlock doesn't clear this entry.
 	}
-	nsLk = n.lockMap[resource]
+	nsLk.ref++
+	n.lockMap[resource] = nsLk
 	n.lockMapMutex.Unlock()
 
 	// Locking here will block (until timeout).
@@ -97,15 +95,19 @@ func (n *nsLockMap) lock(ctx context.Context, volume string, path string, lockSo
 	}
 
 	if !locked { // We failed to get the lock
-
 		// Decrement ref count since we failed to get the lock
-		if atomic.AddUint32(&nsLk.ref, ^uint32(0)) == 0 {
-			// Remove from the map if there are no more references.
-			n.lockMapMutex.Lock()
-			delete(n.lockMap, resource)
-			n.lockMapMutex.Unlock()
+		n.lockMapMutex.Lock()
+		n.lockMap[resource].ref--
+		if n.lockMap[resource].ref < 0 {
+			logger.CriticalIf(GlobalContext, errors.New("resource reference count was lower than 0"))
 		}
+		if n.lockMap[resource].ref == 0 {
+			// Remove from the map if there are no more references.
+			delete(n.lockMap, resource)
+		}
+		n.lockMapMutex.Unlock()
 	}
+
 	return
 }
 
@@ -123,7 +125,11 @@ func (n *nsLockMap) unlock(volume string, path string, readLock bool) {
 	} else {
 		n.lockMap[resource].Unlock()
 	}
-	if atomic.AddUint32(&n.lockMap[resource].ref, ^uint32(0)) == 0 {
+	n.lockMap[resource].ref--
+	if n.lockMap[resource].ref < 0 {
+		logger.CriticalIf(GlobalContext, errors.New("resource reference count was lower than 0"))
+	}
+	if n.lockMap[resource].ref == 0 {
 		// Remove from the map if there are no more references.
 		delete(n.lockMap, resource)
 	}

--- a/cmd/namespace-lock_test.go
+++ b/cmd/namespace-lock_test.go
@@ -17,7 +17,10 @@
 package cmd
 
 import (
+	"context"
+	"runtime"
 	"testing"
+	"time"
 )
 
 // WARNING:
@@ -29,9 +32,69 @@ import (
 func TestGetSource(t *testing.T) {
 	currentSource := func() string { return getSource(2) }
 	gotSource := currentSource()
-	// Hard coded line number, 31, in the "expectedSource" value
-	expectedSource := "[namespace-lock_test.go:31:TestGetSource()]"
+	// Hard coded line number, 34, in the "expectedSource" value
+	expectedSource := "[namespace-lock_test.go:34:TestGetSource()]"
 	if gotSource != expectedSource {
 		t.Errorf("expected : %s, got : %s", expectedSource, gotSource)
+	}
+}
+
+// Test lock race
+func TestNSLockRace(t *testing.T) {
+	ctx := context.Background()
+
+	for i := 0; i < 10000; i++ {
+		nsLk := newNSLock(false)
+
+		// lk1; ref=1
+		if !nsLk.lock(ctx, "volume", "path", "source", "opsID", false, time.Second) {
+			t.Fatal("failed to acquire lock")
+		}
+
+		// lk2
+		lk2ch := make(chan struct{})
+		go func() {
+			defer close(lk2ch)
+			nsLk.lock(ctx, "volume", "path", "source", "opsID", false, 1*time.Millisecond)
+		}()
+		time.Sleep(1 * time.Millisecond) // wait for goroutine to advance; ref=2
+
+		// Unlock the 1st lock; ref=1 after this line
+		nsLk.unlock("volume", "path", false)
+
+		// Taking another lockMapMutex here allows queuing up additional lockers. This should
+		// not be required but makes reproduction much easier.
+		nsLk.lockMapMutex.Lock()
+
+		// lk3 blocks.
+		lk3ch := make(chan bool)
+		go func() {
+			lk3ch <- nsLk.lock(ctx, "volume", "path", "source", "opsID", false, 0)
+		}()
+
+		// lk4, blocks.
+		lk4ch := make(chan bool)
+		go func() {
+			lk4ch <- nsLk.lock(ctx, "volume", "path", "source", "opsID", false, 0)
+		}()
+		runtime.Gosched()
+
+		// unlock the manual lock
+		nsLk.lockMapMutex.Unlock()
+
+		// To trigger the race:
+		// 1) lk3 or lk4 need to advance and increment the ref on the existing resource,
+		//    successfully acquiring the lock.
+		// 2) lk2 then needs to advance and remove the resource from lockMap.
+		// 3) lk3 or lk4 (whichever didn't execute in step 1) then executes and creates
+		//    a new entry in lockMap and acquires a lock for the same resource.
+
+		<-lk2ch
+		lk3ok := <-lk3ch
+		lk4ok := <-lk4ch
+
+		if lk3ok && lk4ok {
+			t.Fatalf("multiple locks acquired; iteration=%d, lk3=%t, lk4=%t", i, lk3ok, lk4ok)
+		}
 	}
 }

